### PR TITLE
DAOS 6524 Server: Compute statistics for gauge and duration metrics

### PR DESCRIPTION
**Pull request by jbrosenz**
Adds statistics computation for the min, max, mean and standard
    deviation for metrics of type duration and gauge.

    Updated tests, example functions, and output to show the
    new statistics.

### CODE REVIEW

**Review by tanabarr — APPROVED** [review]

**Review by kjacque — COMMENTED** [review]
Whoops, it looks like I left all my comments for this changeset on your other PR, #4407. Like I said there, I have a few clarity suggestions, nothing major.

**Comment by jbrosenz** [comment]
I spent some time implementing some macros for the stats computation.  It was functional, but I did not like the results.  I thought it was harder to understand than the original code that didn't use macros.  I decided to remove them because they were not helping in ways that were helpful to me.  I retained the macros for printing, though I think they are not as simple as the original code they replaced.

File: src/gurt/telemetry.c [hunk 1]
```diff
@@ -682,6 +707,101 @@ d_tm_build_path(char **path, char *item, va_list args)
...
+			 (value - mean));
+
+	node->dtn_metric->dtm_stats->mean = mean;
+	node->dtn_metric->dtm_stats->sum_of_squares = sum_of_squares;
+
+	if (node->dtn_metric->dtm_stats->sample_size > 2) {
+		variance = sum_of_squares /
+			   (node->dtn_metric->dtm_stats->sample_size - 1);
+		node->dtn_metric->dtm_stats->std_dev = sqrtl(variance);
+	}
```
Comment (johannlombardi) [inline src/gurt/telemetry.c:747]: It sounds like a lot of calculation on the critical path. I checked the lustre code and they actually only maintain the sumsquare and probably delay the variance calculation to the post processing. Should we do that too and just collect the sum of square to avoid the impact?
Relevant links:
https://github.com/lustre/lustre-release/blob/master/lustre/include/lprocfs_status.h#L153
https://github.com/lustre/lustre-release/blob/master/lustre/obdclass/lprocfs_counters.c#L84

File: src/gurt/telemetry.c [hunk 2]
```diff
@@ -682,6 +707,101 @@ d_tm_build_path(char **path, char *item, va_list args)
...
+
+	if (node->dtn_metric->dtm_stats == NULL)
+		return;
+
+	value = node->dtn_metric->dtm_data.value;
+	node->dtn_metric->dtm_stats->sample_size++;
+
+	mean = node->dtn_metric->dtm_stats->mean +
+	       ((value - node->dtn_metric->dtm_stats->mean) /
+	       node->dtn_metric->dtm_stats->sample_size);
```
Comment (johannlombardi) [inline src/gurt/telemetry.c:734]: Maybe maintaining the sum is enough and we can compute the mean at collect time? Maybe you are worried about overflow?

**Review by tanabarr — DISMISSED** [review]
some of the math is hard to follow and could do with some comments and tests to verify some calculations, otherwise notwithstanding Johann's comment about moving the variance calculation to postprocessing it looks ok

File: src/gurt/telemetry.c [hunk 3]
```diff
@@ -682,6 +707,101 @@ d_tm_build_path(char **path, char *item, va_list args)
...
+ *
+ * \param[in]	node		Pointer to a gauge node
+ */
+void
+d_tm_compute_gauge_stats(struct d_tm_node_t *node)
+{
+	uint64_t	value = 0;
+	double		sum_of_squares = 0.0;
+	double		variance = 0.0;
+	double		mean = 0.0;
```
Comment (johannlombardi) [inline src/gurt/telemetry.c:724]: Maybe define a var on the stack pointing at dtn_metric to make the code more readable?

File: src/gurt/telemetry.c (hunk 1 above)
Comment (jbrosenz) [reply src/gurt/telemetry.c:747]: That's a good idea.  I will go ahead and defer the variance and standard dev calculation until it is read.

File: src/gurt/telemetry.c [hunk 4]
```diff
@@ -1822,14 +1980,22 @@ d_tm_get_duration(struct timespec *tms, uint64_t *shmem_root,
 	if (!d_tm_validate_shmem_ptr(shmem_root, (void *)node))
 		return -DER_METRIC_NOT_FOUND;
 
-	if (!(node->dtn_type & D_TM_DURATION))
+	if (node->dtn_type != D_TM_GAUGE)
 		return -DER_OP_NOT_PERMITTED;
 
 	metric_data = d_tm_conv_ptr(shmem_root, node->dtn_metric);
 	if (metric_data != NULL) {
+
```
[the remark this answers is not in the document]
Comment (jbrosenz) [reply src/gurt/telemetry.c:1988]: fixed

File: src/gurt/telemetry.c (hunk 3 above)
Comment (jbrosenz) [reply src/gurt/telemetry.c:724]: I made a stack var for dtm_stats instead.

File: src/gurt/telemetry.c (hunk 2 above)
Comment (jbrosenz) [reply src/gurt/telemetry.c:734]: The sum of squares needs the old mean and the new mean, so the mean needs to be computed each time a new metric value is generated.

File: src/gurt/telemetry.c [hunk 5]
```diff
@@ -1582,6 +1706,37 @@ d_tm_add_metric(struct d_tm_node_t **node, char *metric, int metric_type,
...
+	temp->dtn_metric->dtm_stats = NULL;
+	if (metric_type == D_TM_GAUGE) {
+		temp->dtn_metric->dtm_stats =
+				     d_tm_shmalloc(sizeof(struct d_tm_stats_t));
+		if (temp->dtn_metric->dtm_stats == NULL) {
+			rc = -DER_NO_SHMEM;
+			goto failure;
+		}
+		temp->dtn_metric->dtm_stats->dtm_min.min_int = UINT64_MAX;
+		temp->dtn_metric->dtm_stats->dtm_max.max_int = 0;
```
[the remark this answers is not in the document]
Comment (jbrosenz) [reply src/gurt/telemetry.c:1718]: The memory for the nodes are allocated in shared memory, obtained from shmget() which does zero init the memory when the memory is allocated but not on subsequent shmget() on existing segments.  However, I can clear the memory when it is allocated from the pool with a memset() and remove the statements that set the individual items to 0.  That's a good idea.  Thanks.

File: src/gurt/tests/test_gurt_telem_producer.c [hunk 6]
```diff
@@ -210,6 +210,99 @@ test_input_validation(void **state)
...
 
+static void
+test_gauge_stats(void **state)
+{
+	static struct d_tm_node_t	*node;
+	int				rc;
+
+	rc = d_tm_set_gauge(&node, 2, "gurt/tests/telem/gauge-stats", NULL);
+	assert(rc == D_TM_SUCCESS);
+	rc = d_tm_set_gauge(&node, 4, NULL);
```
[the remark this answers is not in the document]
Comment (jbrosenz) [reply src/gurt/tests/test_gurt_telem_producer.c:221]: Ok, I un-unrolled the loop and use an array now.  See if that's closer to what you had in mind.

File: src/gurt/telemetry.c [hunk 7]
```diff
@@ -682,6 +707,87 @@ d_tm_build_path(char **path, char *item, va_list args)
...
+
+	value = node->dtn_metric->dtm_data.value;
+	dtm_stats->sample_size++;
+
+	mean = dtm_stats->mean + ((value - dtm_stats->mean) /
+	       dtm_stats->sample_size);
+
+	sum_of_squares = dtm_stats->sum_of_squares +
+			 ((value - dtm_stats->mean) * (value - mean));
+
```
Comment (johannlombardi) [inline src/gurt/telemetry.c:739]: I think that this involves more calculations than:
sum += val;
sum_of_squares += val * val;

File: src/gurt/telemetry.c (hunk 1 above)
Comment (jbrosenz) [reply src/gurt/telemetry.c]: I read the Lustre code today.  I see that they leave the actual standard deviation calculation to some other software.  Interestingly, [https://wiki.lustre.org/Lustre_Monitoring_and_Statistics_Guide](url) says that 

> The standard deviation can be computed from sqrt({sum of values squared} - {mean value}²)

However, that formula is not correct.  It's more accurately: sqrt(({some of values squared} - (sample size * mean²) / (sample size -1)).  See if this new implementation is more to your liking.

File: src/gurt/telemetry.c (hunk 7 above)
Comment (jbrosenz) [reply src/gurt/telemetry.c]: Sure, I've made the changes.  See what you think.

File: src/gurt/telemetry.c [hunk 8]
```diff
@@ -682,6 +707,86 @@ d_tm_build_path(char **path, char *item, va_list args)
...
+	struct d_tm_stats_t	*dtm_stats;
+	double			value = 0;
+
+	dtm_stats = node->dtn_metric->dtm_stats;
+
+	if (dtm_stats == NULL)
+		return;
+
+	value = node->dtn_metric->dtm_data.tms[0].tv_sec +
+		(node->dtn_metric->dtm_data.tms[0].tv_nsec / 1E9);
```
Comment (johannlombardi) [inline src/gurt/telemetry.c:777]: I think that it would be more efficient to store it in nsec and avoid using double and a division here. Can be done in a subsequent patch.

**Review by johannlombardi — DISMISSED** [review]
Looks good. Thanks.

File: src/gurt/telemetry.c (hunk 8 above)
Comment (jbrosenz) [reply src/gurt/telemetry.c:777]: I am not following what you had in mind for storing the value in tv_nsec.  I did try my hand at a fixed point implementation to eliminate the divide within d_tm_compute_duration_stats().  What do you think of this instead?

```

void
d_tm_compute_duration_stats(struct d_tm_node_t *node)
{
	struct d_tm_stats_t	*dtm_stats;
	uint64_t		value = 0;

	dtm_stats = node->dtn_metric->dtm_stats;

	if (dtm_stats == NULL)
		return;

	value = (node->dtn_metric->dtm_data.tms[0].tv_sec * 1E9) +
		 node->dtn_metric->dtm_data.tms[0].tv_nsec;

	dtm_stats->sample_size++;
	dtm_stats->dtm_sum.sum_int += value;
	dtm_stats->sum_of_squares += (long double)value * (long double)value;

	if (value > dtm_stats->dtm_max.max_int)
		dtm_stats->dtm_max.max_int = value;

	if (value < dtm_stats->dtm_min.min_int)
		dtm_stats->dtm_min.min_int = value;
}

```

Then when the data is read, it is converted back to float inside d_tm_get_duration() shown in this snippet:

```

		if (stats != NULL) {
			stats->dtm_min.min_float = (double)dtm_stats->dtm_min.min_int / 1E9;
			stats->dtm_max.max_float = (double)dtm_stats->dtm_max.max_int / 1E9;
			stats->dtm_sum.sum_float = (double)dtm_stats->dtm_sum.sum_int / 1E9;
			if (dtm_stats->sample_size > 0)
				stats->mean = stats->dtm_sum.sum_float /
					      dtm_stats->sample_size;
			stats->sum_of_squares = (double)((dtm_stats->sum_of_squares / 1E18));
			stats->std_dev = d_tm_compute_standard_dev(
						      stats->sum_of_squares,
						      dtm_stats->sample_size,
						      stats->mean);

			stats->sample_size = dtm_stats->sample_size;
		}

```

**Review by tanabarr — APPROVED** [review]
recent changes look good

File: src/gurt/telemetry.c [hunk 9]
```diff
@@ -726,88 +708,83 @@ d_tm_build_path(char **path, char *item, va_list args)
...
+ */
+double
+d_tm_compute_standard_dev(double sum_of_squares, uint64_t sample_size,
+			  double mean)
+{
+	if (sample_size == 0)
+		return 0;
+
+	return sqrtl((sum_of_squares - (sample_size * mean * mean)) /
+		     (sample_size - 1));
```
Comment (kjacque) [inline src/gurt/telemetry.c:727]: What if sample_size == 1?

File: src/gurt/tests/test_gurt_telem_consumer.c [hunk 10]
```diff
@@ -182,6 +190,42 @@ test_timer_snapshot(void **state)
...
+
+	rc = d_tm_get_gauge(&val, &stats, shmem_root, NULL,
+			    "gurt/tests/telem/gauge-stats");
+	assert(rc == D_TM_SUCCESS);
+
+	assert_int_equal(val, 20);
+	assert_int_equal(stats.dtm_min.min_int, 2);
+	assert_int_equal(stats.dtm_max.max_int, 20);
+	assert(stats.mean - 11.0 < STATS_EPSILON);
+	assert(stats.std_dev - 5.89379 < STATS_EPSILON);
```
Comment (kjacque) [inline src/gurt/tests/test_gurt_telem_consumer.c:208]: Did assert_float_equal not work?

**Review by kjacque — CHANGES_REQUESTED** [review]
The divide-by-zero bug should be fixed before landing. The other comment is just a question.

File: src/gurt/telemetry.c (hunk 9 above)
Comment (jbrosenz) [reply src/gurt/telemetry.c:727]: yes, I broke that with this update.  I'll fix it.

File: src/gurt/tests/test_gurt_telem_consumer.c (hunk 10 above)
Comment (jbrosenz) [reply src/gurt/tests/test_gurt_telem_consumer.c:208]: It was not available under all of our builds.  It worked fine locally and for the builds that had the correct version of cmocka that supported it.  I didn't want to expand the scope of this PR to upgrade cmocka at the same time.

File: src/gurt/telemetry.c (hunk 8 above)
Comment (johannlombardi) [reply src/gurt/telemetry.c:809]: Looks good. Integer multiplication calculation is faster than floating point division.

**Review by johannlombardi — APPROVED** [review]

File: src/gurt/tests/test_gurt_telem_consumer.c (hunk 10 above)
Comment (kjacque) [reply src/gurt/tests/test_gurt_telem_consumer.c:208]: Ah, that's fine. Thanks for the explanation.

**Review by kjacque — APPROVED** [review]
